// passport-verify/build/lib/create-response-handler.js
"use strict";
/**
 * The contents of this file are only relevant if your service uses the
 * legacy setup. A legacy setup involves connecting to GOV.UK Verify with a
 * Matching Service Adapter (MSA).
 *
 *
 * create-response-handler
 *
 * Helper function to make handling callbacks from
 * `passport.authenticate('verify', callback)` easier.
 *
 * This is necessary because there are a number of different scenarios
 * that services need to handle when dealing with Verify responses.
 * passport.js isn't aware of these, so the interface it provides is not ideal.
 */
/** */
Object.defineProperty(exports, "__esModule", { value: true });
const translated_response_body_1 = require("./verify-service-provider-api/translated-response-body");
/**
 * The `createResponseHandler` function should be used if your service uses the
 * legacy setup. A legacy setup involves connecting to GOV.UK Verify with a Matching
 * Service Adapter (MSA).
 *
 * This function makes it easier to handle the passport.authenticate() callback. The
 * function takes separate callbacks for each of the response scenarios you have to
 * handle and returns a function that will invoke the appropriate callback when called
 * by passport.
 *
 * @param responseScenarios Callbacks to handle each type of response that GOV.UK Verify can return
 */
function createResponseHandler(responseScenarios) {
    return function (error, user, infoOrError, status) {
        if (error) {
            return responseScenarios.onError(error);
        }
        if (user) {
            const responseBody = infoOrError;
            if (responseBody.scenario === translated_response_body_1.Scenario.ACCOUNT_CREATION) {
                return responseScenarios.onCreateUser(user);
            }
            return responseScenarios.onMatch(user);
        }
        switch (infoOrError) {
            case translated_response_body_1.Scenario.REQUEST_ERROR:
                return responseScenarios.onError(new Error('SAML Response was an error'));
            case translated_response_body_1.Scenario.NO_MATCH:
                return responseScenarios.onNoMatch();
            case translated_response_body_1.Scenario.CANCELLATION:
                return responseScenarios.onCancel();
            case translated_response_body_1.Scenario.AUTHENTICATION_FAILED:
                return responseScenarios.onAuthnFailed();
            default:
                return responseScenarios.onError(new Error(`Unrecognised Scenario ${infoOrError}`));
        }
    };
}
exports.createResponseHandler = createResponseHandler;
